File: Baekjoon/17000~17999/[17204]죽음의 게임/solution.py
from collections import deque


N, M = map(int, input().split())

number = [0 for i in range(N)]
visited = [False for i in range(N)]
q = deque()
res = 1

for i in range(N):
    number[i] = int(input())

q.append(number[0])
visited[0] = True

while len(q) != 0:
    cur = q.popleft()

    if(cur == M):
        break

    if visited[number[cur]] == False:
        q.append(number[cur])
        visited[number[cur]] = True
        res += 1
    else:
        res = -1

print(res)

File: Baekjoon/17000~17999/[17204]죽음의 게임/solution.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

#define endl "\n"

vector<int> number;
vector<bool> visited;
int N, M;

int bfs(int v);

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  cin >> N >> M;

  number = vector<int>(N);
  visited = vector<bool>(N, false);

  for (int i = 0; i < N; i++) {
    cin >> number[i];
  }

  cout << bfs(0) << endl;

  return 0;
}

int bfs(int v) {
  queue<int> q;
  int res = 1;

  q.push(number[v]);
  visited[v] = true;

  while (!q.empty()) {
    int cur = q.front();

    if (cur == M) {
      break;
    }

    q.pop();

    if (!visited[number[cur]]) {
      q.push(number[cur]);
      visited[number[cur]] = true;
      res++;
    } else
      res = -1;
  }

  return res;
}